File: customer-service/src/main/java/com/wks/servicemarketplace/customerservice/config/DatabaseMigration.java
package com.wks.servicemarketplace.customerservice.config;

import com.hubspot.algebra.Result;
import com.wks.servicemarketplace.customerservice.adapters.db.dao.DataSource;
import liquibase.Liquibase;
import liquibase.database.DatabaseFactory;
import liquibase.database.jvm.JdbcConnection;
import liquibase.exception.DatabaseException;
import liquibase.exception.LiquibaseException;
import liquibase.resource.ClassLoaderResourceAccessor;

import javax.inject.Inject;
import java.sql.SQLException;

public class DatabaseMigration {

    private final DataSource dataSource;

    @Inject
    public DatabaseMigration(DataSource dataSource){
        this.dataSource = dataSource;
    }

    public void migrate() {
        try {
            var database = DatabaseFactory.getInstance().findCorrectDatabaseImplementation(new JdbcConnection(dataSource.getConnection()));
            var liquibase = new Liquibase("liquibase/customerService.changelog.xml", new ClassLoaderResourceAccessor(), database);
            liquibase.update("Migration at ${OffsetDateTime.now(Clock.systemUTC())}");
        } catch (Exception e) {
            throw new RuntimeException("Migration Failed", e);
        }
    }
}

File: service-provider-service/src/main/kotlin/com/wks/servicemarketplace/serviceproviderservice/config/DatabaseMigration.kt
package com.wks.servicemarketplace.serviceproviderservice.config

import com.github.michaelbull.result.Result
import com.github.michaelbull.result.mapError
import com.github.michaelbull.result.runCatching
import com.wks.servicemarketplace.serviceproviderservice.adapters.db.dao.DataSource
import liquibase.Liquibase
import liquibase.database.DatabaseFactory
import liquibase.database.jvm.JdbcConnection
import liquibase.exception.DatabaseException
import liquibase.resource.ClassLoaderResourceAccessor
import java.time.Clock
import java.time.OffsetDateTime

sealed class MigrationError(open val message: String?, open val cause: Throwable) {
    data class Database(override val message: String?, override val cause: Throwable) : MigrationError(message, cause)
    data class Migration(override val message: String?, override val cause: Throwable) : MigrationError(message, cause)
}

class DatabaseMigration constructor(private val dataSource: DataSource) {

    fun migrate(): Result<Unit, MigrationError> {
        return runCatching {
            val database =
                DatabaseFactory.getInstance().findCorrectDatabaseImplementation(JdbcConnection(dataSource.connection()))
            val liquibase = Liquibase("liquibase/serviceProviderService.changelog.xml", ClassLoaderResourceAccessor(), database)
            liquibase.update("Migration at ${OffsetDateTime.now(Clock.systemUTC())}")
        }.mapError {
            when(it) {
                is DatabaseException -> MigrationError.Database(it.message, it)
                else -> MigrationError.Migration(it.message, it)
            }
        }
    }
}